src/tensor_operations/matrix_decomposition.jl

```diff
@@ -360,44 +360,92 @@ function noinds_error_message(decomp::String)
 end
 
+function add_trivial_index(A::ITensor, Ainds)
+  α = trivial_index(Ainds) #If Ainds[1] has no QNs makes Index(1), otherwise Index(QN()=>1)
+  vα = onehot(eltype(A), α => 1)
+  A *= vα
+  return A, vα, [α]
+end
+
+function add_trivial_index(A::ITensor, Linds, Rinds)
+  vαl, vαr = nothing, nothing
+  if isempty(Linds)
+    A, vαl, Linds = add_trivial_index(A, Rinds)
+  end
+  if isempty(Rinds)
+    A, vαr, Rinds = add_trivial_index(A, Linds)
+  end
+  return A, vαl, vαr, Linds, Rinds
+end
+
+function remove_trivial_index(Q::ITensor, R::ITensor, vαl, vαr)
+  if !isnothing(vαl) #should have only dummy + qr,Link
+    Q *= dag(vαl)
+  end
+  if !isnothing(vαr) #should have only dummy + qr,Link
+    R *= dag(vαr)
+  end
+  return Q, R
+end
+
 qr(A::ITensor; kwargs...) = error(noinds_error_message("qr"))
 
-# TODO: write this in terms of combiners and then
-# call qr on the order-2 tensors directly
 function qr(A::ITensor, Linds...; kwargs...)
-  tags::TagSet = get(kwargs, :tags, "Link,qr")
+  qtag::TagSet = get(kwargs, :tags, "Link,qr") #tag for new index between Q and R
   Lis = commoninds(A, indices(Linds...))
   Ris = uniqueinds(A, Lis)
-
-  Lis_original = Lis
-  Ris_original = Ris
-  if isempty(Lis_original)
-    α = trivial_index(Ris)
-    vLα = onehot(eltype(A), α => 1)
-    A *= vLα
-    Lis = [α]
-  end
-  if isempty(Ris_original)
-    α = trivial_index(Lis)
-    vRα = onehot(eltype(A), α => 1)
-    A *= vRα
-    Ris = [α]
+  lre = isempty(Lis) || isempty(Ris)
+  # make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  if (lre)
+    A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
   end
 
-  Lpos, Rpos = NDTensors.getperms(inds(A), Lis, Ris)
-  QT, RT = qr(tensor(A), Lpos, Rpos; kwargs...)
-  Q, R = itensor(QT), itensor(RT)
-  q = commonind(Q, R)
-  settags!(Q, tags, q)
-  settags!(R, tags, q)
-  q = settags(q, tags)
-
-  if isempty(Lis_original)
-    Q *= dag(vLα)
+  #
+  #  Use combiners to render A down to a rank 2 tensor ready matrix QR routine.
+  #
+  CL, CR = combiner(Lis...), combiner(Ris...)
+  cL, cR = combinedind(CL), combinedind(CR)
+  AC = A * CR * CL
+  #
+  #  Make sure we don't accidentally pass the transpose into the matrix qr routine.
+  #
+  if inds(AC) != IndexSet(cL, cR)
+    AC = permute(AC, cL, cR)
   end
-  if isempty(Ris_original)
-    R *= dag(vRα)
+  # qr the matrix.
+  QT, RT = qr(tensor(AC); kwargs...)
+
+  # We need a use case that fails without this code.
+  # correcting the fluxes of the two tensors, such that Q has 0 flux for all blocks
+  # and R has the total flux of the system
+  # if hasqns(AC)
+  # for b in nzblocks(QT)
+  #     i1 = inds(QT)[1]
+  #     i2 = inds(QT)[2]
+  #     r1 = inds(RT)[1]
+  #     newqn = -dir(i2) * flux(i1 => Block(b[1]))
+  #     ITensors.setblockqn!(i2, newqn, b[2])
+  #     ITensors.setblockqn!(r1, newqn, b[2])
+  # end
+  # end
```

Comment:
For now we can just delete this and fix any bugs properly in NDTensors if they come up.

Reply:
Actually I have un-commented it, put a mirror version in rq() and updated the tests to verify zero flux between Q&R when the input has zero flux.  This ends up being important for orthogonalization of certain MPOs.

Reply:
Could we please move comments to #1012 if that is where the QR code will be developed? It is confusing tracking two PRs.